proto/pulumi/provider.proto

```diff
@@ -115,13 +115,23 @@ service ResourceProvider {
     // Call dynamically executes a method in the provider associated with a component resource.
     rpc Call(CallRequest) returns (CallResponse) {}
 
-    // Check validates that the given property bag is valid for a resource of the given type and returns the inputs
-    // that should be passed to successive calls to Diff, Create, or Update for this resource. As a rule, the provider
-    // inputs returned by a call to Check should preserve the original representation of the properties as present in
-    // the program inputs. Though this rule is not required for correctness, violations thereof can negatively impact
-    // the end-user experience, as the provider inputs are using for detecting and rendering diffs.
+    // `Check` validates a set of input properties against a given resource type. A `Check` call returns either a set of
+    // checked, known-valid inputs that may subsequently be passed to [](pulumirpc.ResourceProvider.Diff),
+    // [](pulumirpc.ResourceProvider.Create), or [](pulumirpc.ResourceProvider.Update); or a set of errors explaining
+    // why the inputs are invalid. In the case that a set of inputs are successfully validated and returned, `Check`
+    // *may also populate default values* for resource inputs, returning them so that they may be passed to a subsequent
+    // call and persisted in the Pulumi state. In the case that `Check` fails and returns a set of errors, it is
+    // expected that the caller (typically the Pulumi engine) will fail resource registration.
+    //
+    // As a rule, the provider inputs returned by a call to `Check` should preserve the original representation of the
+    // properties as present in the program inputs. Though this rule is not required for correctness, violations thereof
+    // can negatively impact the end-user experience, as the provider inputs are using for detecting and rendering
+    // diffs.
     rpc Check(CheckRequest) returns (CheckResponse) {}
-    // Diff checks what impacts a hypothetical update will have on the resource's properties.
+
+    // `Diff` compares an existing ("old") set of resource properties with a new set of properties and computes the
```

Comment:
Note that olds might require migrating to the current schema.

Reply:
Note also that aliasing may cause the olds to come from an aliased resource, even one with a different type.  How would one determine the type of the olds?